### PR TITLE
Make the tree node class configurable.

### PR DESCRIPTION
**Pull request by Ner1Co**

### CODE REVIEW

File: dist/TreeModel.js [hunk 1]
```diff
@@ -19,8 +19,13 @@ module.exports = (function () {
   function TreeModel(config) {
     config = config || {};
     this.config = config;
+    this.config.nodeClass = config.nodeClass || Node;
     this.config.childrenPropertyName = config.childrenPropertyName || 'children';
     this.config.modelComparatorFn = config.modelComparatorFn;
+
+    if (!(Node.isPrototypeOf(this.config.nodeClass) || this.config.nodeClass === Node)) {
```
Comment (byohay) [inline dist/TreeModel.js]: IMO `!Node.isPrototypeOf(this.config.nodeClass) && this.config.nodeClass !== Node` is more readable.

File: index.js [hunk 2]
```diff
@@ -18,8 +18,13 @@ module.exports = (function () {
   function TreeModel(config) {
     config = config || {};
     this.config = config;
+    this.config.nodeClass = config.nodeClass || Node;
     this.config.childrenPropertyName = config.childrenPropertyName || 'children';
     this.config.modelComparatorFn = config.modelComparatorFn;
+
+    if (!(Node.isPrototypeOf(this.config.nodeClass) || this.config.nodeClass === Node)) {
+      throw new TypeError('config.nodeClass must be of type Node or extended Node class.');
```
Comment (byohay) [inline index.js]: Same.

File: test/test.js [hunk 3]
```diff
@@ -814,4 +814,25 @@ describe('TreeModel', function () {
...
+    var treeModel;
+
+    it('should throw an error when node class is invalid', function () {
+      function nodeClass() {}
+
+      assert.throws(function(){new TreeModel({nodeClass: nodeClass})}, TypeError,
+          'config.nodeClass must be of type Node or extended Node class.');
+    });
+
+    it('should config and parse without errors', function () {
```
Comment (byohay) [inline test/test.js]: The main test here is checking that the tree is itself a `CustomNodeClass` class, so change the title accordingly.

File: test/test.js [hunk 4]
```diff
@@ -814,4 +814,25 @@ describe('TreeModel', function () {
...
+      function CustomNodeClass() {}
+      CustomNodeClass.__proto__ = TreeModel.Node;
+
+      treeModel = new TreeModel({nodeClass: CustomNodeClass});
+      var tree = treeModel.parse({node: "info"});
+
+      assert.equal(tree instanceof CustomNodeClass, true);
```
Comment (byohay) [inline test/test.js]: Just `assert(tree instanceof CustomNodeClass);`

File: test/test.js [hunk 5]
```diff
@@ -814,4 +814,25 @@ describe('TreeModel', function () {
...
+    it('should throw an error when node class is invalid', function () {
+      function nodeClass() {}
+
+      assert.throws(function(){new TreeModel({nodeClass: nodeClass})}, TypeError,
+          'config.nodeClass must be of type Node or extended Node class.');
+    });
+
+    it('should config and parse without errors', function () {
+      function CustomNodeClass() {}
+      CustomNodeClass.__proto__ = TreeModel.Node;
```
Comment (byohay) [inline test/test.js]: Maybe you can define that `CustomNodeClass` extends `Node`?

File: test/test.js (hunk 4 above)
Comment (byohay) [reply test/test.js]: Why not using `isPrototypeOf` here? (I'm because I don't know what's better).

File: test/test.js [hunk 6]
```diff
@@ -814,4 +814,25 @@ describe('TreeModel', function () {
...
+
+      treeModel = new TreeModel({nodeClass: CustomNodeClass});
+      var tree = treeModel.parse({node: "info"});
```
Comment (byohay) [inline test/test.js]: We should continue with the same variable names, e.g. `id`.

File: README.md [hunk 7]
```diff
@@ -41,6 +41,7 @@ Valid properties for the options object are:
 
 * `childrenPropertyName` - The name for the children array property. Default is `children`;
 * `modelComparatorFn` - A comparator function to sort the children when parsing the model and adding children. The default order policy is to keep the parsed order and append new children. The comparator function receives the model for two nodes just like the [Array.sort](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) function. The provided sort algorithm is **stable**.
+* `nodeClass` - The class that is using for creating the tree nodes, must extend `Node` -  otherwise error is thrown. Default is `Node` itself;
```
Comment (byohay) [inline README.md]: `that is used for creating the tree nodes. Must...`.
`otherwise an error...`

File: dist/TreeModel.js [hunk 8]
```diff
@@ -288,6 +293,7 @@ module.exports = (function () {
     return this;
   };
 
+  TreeModel.Node = Node;
```
Comment (byohay) [inline dist/TreeModel.js:296]: What's this for? Don't you need `TreeModel.Node = this.config.nodeClass`?

**Review by byohay — CHANGES_REQUESTED** [review]
Round

File: test/test.js (hunk 5 above)
Comment (Ner1Co) [reply test/test.js]: this is the dirty way since it's not ES6 here..

File: dist/TreeModel.js (hunk 1 above)
Comment (Ner1Co) [reply dist/TreeModel.js]: Done.

File: dist/TreeModel.js (hunk 8 above)
Comment (Ner1Co) [reply dist/TreeModel.js:296]: It's for exposing the class Node outside so we can extend it.

File: index.js (hunk 2 above)
Comment (Ner1Co) [reply index.js]: Done.

File: test/test.js (hunk 3 above)
Comment (Ner1Co) [reply test/test.js]: Done.

File: test/test.js (hunk 6 above)
Comment (Ner1Co) [reply test/test.js]: Done.

File: README.md (hunk 7 above)
Comment (Ner1Co) [reply README.md]: Done.

File: test/test.js [hunk 9]
```diff
@@ -814,4 +814,26 @@ describe('TreeModel', function () {
...
     });
   });
+  
+  describe('with custom node class', function () {
+    var treeModel;
+
+    it('should throw an error when node class is invalid', function () {
+      function nodeClass() {}
+
+      assert.throws(function(){new TreeModel({nodeClass: nodeClass})}, TypeError,
```
Comment (byohay) [inline test/test.js]: 1. The string is the message that is displayed when the assertion fails, not asserting what the message should be. So just remove it.
2. IMO this is more readable:
```js
      assert.throws(
          function () {
            new TreeModel({nodeClass: nodeClass})
          },
          TypeError
      );
```

File: test/test.js [hunk 10]
```diff
@@ -814,4 +814,26 @@ describe('TreeModel', function () {
...
+
+    it('should create nodes by the custom class', function () {
+      function CustomNodeClass() {}
+      CustomNodeClass.__proto__ = TreeModel.Node;
+
+      treeModel = new TreeModel({nodeClass: CustomNodeClass});
+      var tree = treeModel.parse({id: "node"});
+
+      assert.ok(tree instanceof CustomNodeClass);
+
```
Comment (byohay) [inline test/test.js]: Remove.

File: test/test.js [hunk 11]
```diff
@@ -814,4 +814,26 @@ describe('TreeModel', function () {
...
+    });
+
+    it('should create nodes by the custom class', function () {
+      function CustomNodeClass() {}
+      CustomNodeClass.__proto__ = TreeModel.Node;
+
+      treeModel = new TreeModel({nodeClass: CustomNodeClass});
+      var tree = treeModel.parse({id: "node"});
+
+      assert.ok(tree instanceof CustomNodeClass);
```
Comment (byohay) [inline test/test.js]: Just `assert` works for me.

**Review by byohay — CHANGES_REQUESTED** [review]
F&S

File: test/test.js (hunk 10 above)
Comment (Ner1Co) [reply test/test.js]: Done.

File: test/test.js (hunk 11 above)
Comment (Ner1Co) [reply test/test.js]: Done.

File: test/test.js (hunk 9 above)
Comment (Ner1Co) [reply test/test.js]: Done.

**Review by byohay — APPROVED** [review]

**Pull request merged**